### ISSUE TITLE
Nameplate / Cutsheet / Scalar Properties of Brick Entities

### ISSUE THREAD

**Issue by gtfierro** [BrickSchema/Brick#189]
Need to discuss how to model these kinds of scalar properties, which may be associated with specific types/makes/models of equipment, but due to scalar values may not make sense as a new subclass.

Location examples:
- sq footage (gross, net)
- volume
- R factor

Equipment examples:
- max rated power/speed
- mode sensitive limits

**We should begin bootstrapping the list of properties that we would like to add to Brick**.  We will also need to decide how this will be modeled. Some options include the following:

```ttl
# option 1
:room1  a  brick:Room ;
       prop:grossSquareArea   "200"^^square_feet .

# option 2
:room2  a brick:Room ;
   brick:hasProperty  :room2grossarea .
:room2grossarea   a   prop:grossSquareArea ;
    prop:hasValue  "200"^^xsd:integer ;
    prop:hasUnit  qudt:SquareFoot .
```

We will need to strike a balance between pushing complexity into the Brick model and complexity in the query model

**Comment by gtfierro** [comment]
An alternative design:

```ttl
:room3 a brick:Room ;
  brick:hasGrossSquareArea :room3grossarea .
:room3grossarea a brick:AreaQuantity ;
      prop:hasValue  "200"^^xsd:integer ;
      prop:hasUnit  qudt:SquareFoot .
```

which could also be written as

```ttl
:room3 a brick:Room ;
  brick:hasGrossSquareArea [
      a brick:AreaQuantity ;
      prop:hasValue  "200"^^xsd:integer ;
      prop:hasUnit  qudt:SquareFoot 
] .
```

**Comment by Reapor-Yurnero** [comment]
Here's a list of scalar properties that I can think of so far:

Brick:Room <- Area, Height, Capacity, Volume(somehow redundant but might be valuable in some sense)Brick:VAV (probably also other terminal units), Brick:Fan, Brick:AHU <- Maximum CFMBrick:Equipment <- Rate Power (mostly applicable)Brick:Filter <- CADR rating (for indoor room-level air-purifier)
For most Brick:Equipment, I believe there exist many other useful scalar properties to associate with each of them but more domain expertise is required.

**Comment by epaulson** [comment]
I'm not a fan of option 2, and would rather us start with option 1, sticking with a set of scalar properties to add into our class hierarchy. It seems like that's on the path towards option 3 (which Brick arguably already has - like I think you could write this today:
```ttl
:floor1 a brick:Floor;
    brick:hasPart [
       a brick:Room;
       brick:isFedBy :ahu1;
    ]
```
There's just not a lot interesting things in Brick that you can say and put into that blank node today, but you could say other things:
```ttl
@prefix schema: <https://schema.org/> .
:floor1 a brick:Floor;
    brick:hasPart [
       a brick:Room;
       brick:isFedBy :ahu1;
       schema:floorSize [
           a schema:QuantitativeValue ;
           schema:unitCode "MTK";
           schema:value "25";
       ];
    ]
```

Turns out Jason was thinking about this a few years ago, see https://github.com/schemaorg/schemaorg/issues/1390 

So long as we're careful about adding new properties to the schema I think we're OK. Wikidata is at 91 million items and is really loose about adding to that, but adding a new property is like pulling teeth and they probably need to be strict there. (Though they're already at like 9,000 properties)

**Comment by gtfierro** [comment]
I like Option 3 the best as well: we can introduce a principled set of ObjectProperties and DataProperties that convey the semantic relationship between the entity and the property itself. If the property is itself a blank node, then it can easily have additional metadata properties attached.

Cool to see the old discussion on the schema org repo. The ontology design points are fairly well known; I think the key is determining how much expressive power we need and balancing that with the potential query/model complexity. I also think we should go with QUDT rather than schema.org so that we are consistent with Brick's use of QUDT so far as well as ASHRAE 223P.

Here's a cut at how this would look for floor geometry:

```ttl
:floor1 a   brick:Floor ;
    brick:hasGeometry [
        a   brick:GrossArea ;
        brick:hasUnit qudt:FT2 ; # squared foot
        qudt:value  "2200"^^xsd:integer
    ] ,
    [
        a   brick:NetArea ;
        brick:hasUnit qudt:FT2 ; # squared foot
        qudt:value  "1900"^^xsd:integer
    ] ,
    [
        a   brick:Volume ;
        brick:hasUnit qudt:FT3 ; # cubed foot
        qudt:value  "22000"^^xsd:integer
    ] .
```

What I like here is:
- the relationship `hasGeometry` is generic to the exact quantitykind of the properties that are attached; this allows us to extend the kinds of Brick quantitykinds that we support. If we need to add a new kind of geometry quantity kind, it can easily fit into the existing structure that contains GrossArea, NetArea, Volume, etc.
- we can associate properties with the quantities themselves, e.g. unit, value and perhaps eventually measurementInterval, type of aggregation, etc
- this can reduce the number of relationships that we need to cover our bases: `hasGeometry`, `hasIdentifier`, etc might be possibilities.

**Comment by jbkoh** [comment]
I like the approach of having blank nodes to represent scalar quantities in general. However, I'm not sure if the intermediate properties are usable. Basically we are putting what the value would mean in two places as the property (`hasGeometry`) and the value's class (`GrossArea`). Two issues get me concerned.

- Those intermediate properties became redundant, which would reduce the usability and may cause confusion to end users.
- Would there be always a proper intermediate property for all the scalar properties? For example, I can't think of a good one for maximum capacities of rooms. 

In summary, I would like to model the scala property type only in a single place (instead of both property and scalar type in the last proposal.) For example,
```turtle
:room3 a brick:Room ;
  brick:grossSquareArea [   # Revmoing "has" to show to distinguish it from relationships between entities.
      a brick:ScalarQuantity ; # Or just remove this one.
      prop:hasValue  "200"^^xsd:integer ;
      prop:hasUnit  qudt:SquareFoot 
] .
```

**Comment by gtfierro** [comment]
Note that the use of blank nodes is just a convenient notation; these could easily be first-class NamedIndividuals as well.

In the current model, the property type (I think this is what you mean by "intermediate property"?), e.g. `brick:GrossArea` would likely be an instance of `brick:Quantity` in order to line up with how our Sensor types measure instances of `brick:Quantity` (e.g. a temperature sensor measures `brick:Temperature`, which aligns with QUDT's quantity kinds). QuantityKinds can be dimensionless, which could be used for things like counting (which would be appropriate for the max capacity of the room)

It is definitely an option to put more typing information into the property name. There are two ways to achieve this:

- via `rdfs:subPropertyOf`; this means that any subproperty (e.g. `brick:grossSquareArea`) also inherit parent properties (e.g. `brick:area` and `brick:geometry`), if we cared to organize these into a hierarchy
- if we do *not* have a hierarchy, then we end up with a flat namespace of properties. The lack of organization of the properties means that it might be harder to find the property names that are appropriate for what you are doing

If we don't include the property type for each, then we will need to put the annotations for appropriate units and quantitykind *somewhere*, probably in the `rdfs:range` or a SHACL shape for the property name.

**Comment by jbkoh** [comment]
Yeah, we can introduce the hierarchy if it turns out necessary.

Another alternative to to keep the `Quantity` structure, we can also do `hasPoint` or `hasScalarProperty` and then have `GrossArea` quantity values like
```turtle
:floor1 a   brick:Floor ;
    brick:hasScalarQuantity [
        a   brick:GrossArea ;
        brick:hasUnit qudt:FT2 ; # squared foot
        qudt:value  "2200"^^xsd:integer
    ] ,
```

I don't have a strong opinion on either one over the other, but would like to avoid redundancy. (Of course, either subclasses or subproperties hierarchies are fine.)

**Comment by gtfierro** [comment]
In the interest of moving this forward, I took a stab at a prototype to see how it might look. I decided to introduce a hierarchy of `EntityProperty` classes, related to Brick entities with a `brick:hasProperty` property --- this is how a similar structure is being defined in 223P so it is in our interest to take a similar approach.

The hierarchy is thus. The immediate subclasses are:
- `PhysicalProperty`: A property of the entity's physical existence
- `VirtualProperty`: A property of the entity's virtual (networked, computed) existence
- `IdentityProperty`: A property that identifies the entity in some way
- `InherentProperty`: A permanent, characteristic property of the entity
- `ConfigurableProperty`: A temporary, changeable property of the entity

The first 3 are one "dimension", and the last 2 are the other "dimension". This gives us 6 further subclasses for organizing our properties:
- `InherentPhysicalProperty`
- `ConfigurablePhysicalProperty`
- `InherentVirtualProperty`
- `ConfigurableVirtualProperty`
- `InherentIdentityProperty`
- `ConfigurableIdentityProperty`

The implementation is [here](https://gist.github.com/gtfierro/724242dde0ed504c1c5c471ebe04fcd1). Eventually, I'd like to swap out the literals like `"net"` and `"active"` with actual named individuals with IRIs and real definitions. Also TODO is inferring the valid units based on properties like `prop:PowerComplexity = "active"`.

One note is that the current implementation does make use of SHACL inferencing (which should be available in pySHACL); it's not clear to me yet how the inference would factor in to general use

**Comment by epaulson** [comment]
I think I need to learn more SHAC, and you note that this is still an open question but I'm nervous about adding in SHACL inferencing. Lots of triple stores support OWL inference rules - is SHACL well-supported? 

On Virtual properties - it's common for some quantities to be measured physically and othertimes be computed, and for that to be different depending on the meter - like for the same property one meter might measure it directly and a different meter might compute it. Maybe I'm reading too much into the intent of VirtualProperty, and I'm being thrown by the prototype is saying electricalProperty is a subclass of prop:InherentPhysicalProperty - in this case, even if my meter is computing some property, is it still OK that the property is normally a "physical" property?

Also did you decide to go with option 2?

**Comment by gtfierro** [comment]
We can likely express many of the SHACL constructs I'm using here with OWL (I'm not using anything too fancy), but I do think that the SHACL expressions are easier to read than their equivalents in OWL.

I'm also open to rejiggering the organization of the EntityProperty classes. It all comes down to what we think a decent organization is: maybe virtual properties should only be properties of the virtual representation of an entity, and the "computed" aspect of that gets thrown out. Then, all ElectricalProperties, regardless of whether they are measured or computed, would be InherentPhysicalProperties.

The intent of the design was to be closer to Option 3, but I decided to see how it would work with a hierarchy of Property types, rather than descriptive relationships (because this is closer to what 223P has). I believe we could accomplish a similar structure using `prop:hasInherentPhysicalProperty` as a superproperty of `prop:hasPhaseCount` and so on. We could also consider having these descriptive relationships inferred from the Property types...

**Comment by epaulson** [comment]
Option 2 is like extreme database normalization, and I think errs a bit too far away from developer-friendliness. I'm not dead-set against it, it just feels like it needs a lot of extra hops in the query.

**Comment by gtfierro** [comment]
I agree that we should probably choose a design that simplifies the queries at the expense of model complexity.

Related to this, I've been thinking that we might want to push the semantics into the relationship rather than into the property type. 

```ttl
:floor1 a   brick:Floor ;
    brick:hasGrossArea [
        a   qudtqk:Area ;  # inferred automatically
        brick:hasUnit qudt:FT2 ; # squared foot
        qudt:value  "2200"^^xsd:integer
    ] 
```

and then queries can look like

```sparql
SELECT ?area ?floor WHERE {
  ?floor  a brick:Floor .
  ?floor brick:hasGrossArea/qudt:value ?area
}
```
So the class hierarchy of property types in my prototype would turn into a hierarchy of *relationships*. The values themselves are still blank nodes with some annotations, but the types themselves would follow the general QUDT structure
